kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_with_constraints.h

```diff
@@ -0,0 +1,1288 @@
+
+        KRATOS_INFO_IF("ResidualBasedEliminationBuilderAndSolverWithConstraints", (this->GetEchoLevel() >= 1 && rModelPart.GetCommunicator().MyPID() == 0)) << "System solve time: " << stop_solve - start_solve << std::endl;
+
+        KRATOS_INFO_IF("ResidualBasedEliminationBuilderAndSolverWithConstraints", (this->GetEchoLevel() == 3)) <<
+        "After the solution of the system" << "\nSystem Matrix = " << rA << "\nUnknowns vector = " << rDx << "\nRHS vector = " << rb << std::endl;
+
+        KRATOS_CATCH("")
+    }
+
+    /**
+     * @brief This function is exactly same as the ConstructMatrixStructure() function in base class except that the function
+     * @details Has the call to ApplyConstraints function call once the element and conditions compute their equation ids
+     * @todo Move this method to a common class with block builder and solver with constraints
+    */
+    virtual void ConstructMatrixStructureWithConstraints(
+        typename TSchemeType::Pointer pScheme,
+        TSystemMatrixType& rA,
+        ModelPart& rModelPart
+        )
+    {
+        // Filling with zero the matrix (creating the structure)
+        Timer::Start("MatrixStructure");
+
+        // Getting the elements from the model
+        const int number_of_elements = static_cast<int>(rModelPart.Elements().size());
+
+        // Getting the array of the conditions
+        const int number_of_conditions = static_cast<int>(rModelPart.Conditions().size());
+
+        ProcessInfo& r_current_process_info = rModelPart.GetProcessInfo();
+        auto el_begin = rModelPart.ElementsBegin();
+        auto cond_begin = rModelPart.ConditionsBegin();
+
+        const SizeType equation_size = BaseType::mEquationSystemSize;
+
+        std::vector<IndexSetType> indices(equation_size);
+
+        #pragma omp parallel for firstprivate(equation_size)
+        for (int index = 0; index < static_cast<int>(equation_size); ++index)
+            indices[index].reserve(40);
+
+        /// Definition of the eqautio id vector type
+        EquationIdVectorType ids(3, 0);
+
+        #pragma omp parallel for firstprivate(number_of_elements, ids)
+        for (int i_elem = 0; i_elem<number_of_elements; i_elem++) {
+            auto it_elem = el_begin + i_elem;
+            pScheme->EquationId( *(it_elem.base()), ids, r_current_process_info);
+
+            for (IndexType i = 0; i < ids.size(); ++i) {
+                if (ids[i] < BaseType::mEquationSystemSize) {
+                #ifdef _OPENMP
+                    omp_set_lock(&BaseType::mlock_array[ids[i]]);
+                #endif
+                    auto& row_indices = indices[ids[i]];
+                    for (auto it = ids.begin(); it != ids.end(); ++it) {
+                        if (*it < BaseType::mEquationSystemSize)
+                            row_indices.insert(*it);
+                    }
+                #ifdef _OPENMP
+                    omp_unset_lock(&BaseType::mlock_array[ids[i]]);
```

Comment:
if you really need the array of locks for this operation, define it in this function and destroy on exit